crates/sui-analytics-indexer/src/analytics_processor.rs

```diff
@@ -29,72 +29,73 @@ use crate::{
     EPOCH_DIR_PREFIX,
 };
 
-pub struct AnalyticsProcessor<S: Serialize + ParquetSchema> {
-    handler: Box<dyn AnalyticsHandler<S>>,
-    writer: Box<dyn AnalyticsWriter<S>>,
+struct State<S: Serialize + ParquetSchema> {
     current_epoch: u64,
     current_checkpoint_range: Range<u64>,
     last_commit_instant: Instant,
+    num_checkpoint_iterations: u64,
+    writer: Box<dyn AnalyticsWriter<S>>,
+}
+
+pub struct AnalyticsProcessor<S: Serialize + ParquetSchema> {
+    handler: Box<dyn AnalyticsHandler<S>>,
+    state: Mutex<State<S>>,
     metrics: AnalyticsMetrics,
     config: AnalyticsIndexerConfig,
     sender: mpsc::Sender<FileMetadata>,
     #[allow(dead_code)]
     kill_sender: oneshot::Sender<()>,
     #[allow(dead_code)]
     max_checkpoint_sender: oneshot::Sender<()>,
-    num_checkpoint_iterations: u64,
 }
 
 const CHECK_FILE_SIZE_ITERATION_CYCLE: u64 = 50;
 
 #[async_trait::async_trait]
-impl<S: Serialize + ParquetSchema + 'static> Handler for AnalyticsProcessor<S> {
-    fn name(&self) -> &str {
-        self.handler.name()
-    }
-
-    async fn process_checkpoint(&mut self, checkpoint_data: &CheckpointData) -> Result<()> {
+impl<S: Serialize + ParquetSchema + 'static> Worker for AnalyticsProcessor<S> {
+    async fn process_checkpoint(&self, checkpoint_data: CheckpointData) -> Result<()> {
         // get epoch id, checkpoint sequence number and timestamp, those are important
         // indexes when operating on data
         let epoch: u64 = checkpoint_data.checkpoint_summary.epoch();
         let checkpoint_num: u64 = *checkpoint_data.checkpoint_summary.sequence_number();
         let timestamp: u64 = checkpoint_data.checkpoint_summary.data().timestamp_ms;
         info!("Processing checkpoint {checkpoint_num}, epoch {epoch}, timestamp {timestamp}");
-        if epoch > self.current_epoch {
-            self.cut().await?;
-            self.update_to_next_epoch(epoch);
-            self.create_epoch_dirs()?;
-            self.reset()?;
+        let mut state = self.state.lock().await;
```

Comment:
just curious: why does the state need to be locked if it can only be processed one at a time

Reply:
so although this specific case has a concurrency of 1, the framework doesn't make exceptions and requires workers to be thread-safe. So they could be canceled, cloned, and rescheduled in another thread. We do the same for the archival workflow